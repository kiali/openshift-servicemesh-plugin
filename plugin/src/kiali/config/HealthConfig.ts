import { HealthConfig, RegexConfig } from '../types/ServerConfig';

const allMatch = new RegExp('.*');

/*
 Parse configuration from backend format to regex expression
*/
export const parseHealthConfig = (healthConfig: HealthConfig) => {
  for (let [key, r] of Object.entries(healthConfig.rate)) {
    healthConfig.rate[key].namespace = getExpr(healthConfig.rate[key].namespace);
    healthConfig.rate[key].name = getExpr(healthConfig.rate[key].name);
    healthConfig.rate[key].kind = getExpr(healthConfig.rate[key].kind);
    for (let t of Object.values(r.tolerance)) {
      t.code = getExpr(t.code, true);
      t.direction = getExpr(t.direction);
      t.protocol = getExpr(t.protocol);
    }
  }
  return healthConfig;
};

/*
  Convert the string to regex, if isCode is true then call to replaceXCode to change the X|x in code expression to \d
*/
export const getExpr = (value: RegexConfig | undefined, isCode: boolean = false): RegExp => {
  if (value) {
    if (typeof value === 'string' && value !== '') {
      const v = value.replace('\\\\', '\\');
      return new RegExp(isCode ? replaceXCode(v) : v);
    }
    if (typeof value === 'object' && value.toString() !== '/(?:)/') {
      return value;
    }
  }
  return allMatch;
};

/* Replace x|X by the regular expression
   Example: 4XX or 5XX to 4\d\d 5\d\d
*/
const replaceXCode = (value: string): string => {
  return value.replace(/x|X/g, '\\d');
};

/*
 Export for tests
*/
export const allMatchTEST = allMatch;
export const getExprTEST = getExpr;
export const replaceXCodeTEST = replaceXCode;
