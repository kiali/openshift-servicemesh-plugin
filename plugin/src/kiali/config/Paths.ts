/**
 * Paths
 */

export enum Paths {
  AGGREGATES = 'aggregates',
  APPLICATIONS = 'applications',
  ISTIO = 'istio',
  TRACING = 'tracing',
  MESH = 'mesh',
  SERVICES = 'services',
  SERVICEENTRIES = 'istio/serviceentries',
  WORKLOADS = 'workloads'
}
