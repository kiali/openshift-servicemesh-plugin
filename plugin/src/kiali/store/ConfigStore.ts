import { createStore, applyMiddleware, compose, Store } from 'redux';
import { KialiAppState } from './Store';
import { persistStore, persistReducer, Transform } from 'redux-persist';
import { persistFilter } from 'redux-persist-transform-filter';
import { createTransform } from 'redux-persist';

import { rootReducer } from '../reducers';
import thunk from 'redux-thunk';

// defaults to localStorage for web and AsyncStorage for react-native
import storage from 'redux-persist/lib/storage';
import { INITIAL_GLOBAL_STATE } from '../reducers/GlobalState';
import { INITIAL_LOGIN_STATE } from '../reducers/LoginState';
import { INITIAL_GRAPH_STATE } from '../reducers/GraphDataState';
import { INITIAL_USER_SETTINGS_STATE } from '../reducers/UserSettingsState';
import { INITIAL_MESSAGE_CENTER_STATE } from '../reducers/MessageCenter';
import { INITIAL_STATUS_STATE } from '../reducers/HelpDropdownState';
import { INITIAL_NAMESPACE_STATE } from '../reducers/NamespaceState';
import { INITIAL_CLUSTER_STATE } from '../reducers/ClusterState';
import { INITIAL_TRACING_STATE } from '../reducers/TracingState';
import { INITIAL_MESH_TLS_STATE } from '../reducers/MeshTlsState';
import { INITIAL_TOUR_STATE } from '../reducers/TourState';
import { INITIAL_ISTIO_STATUS_STATE } from '../reducers/IstioStatusState';
import { INITIAL_METRICS_STATS_STATE } from '../reducers/MetricsStatsState';
import { INITIAL_ISTIO_CERTS_INFO_STATE } from 'reducers/IstioCertsInfoState';
import { KialiAppAction } from 'actions/KialiAppAction';
import { INITIAL_MESH_STATE } from 'reducers/MeshDataState';
import { webRoot } from 'app/History';

declare const window;

const persistKey = `kiali-${webRoot !== '/' ? webRoot.substring(1) : 'root'}`;

// Needed to be able to whitelist fields but allowing to keep an initialState
const whitelistInputWithInitialState = (
  reducerName: string,
  inboundPaths: string[],
  initialState: any
): Transform<unknown, Transform<unknown, unknown>> =>
  createTransform(
    inboundState => persistFilter(inboundState, inboundPaths, 'whitelist'),
    outboundState => ({ ...initialState, ...outboundState }),
    { whitelist: [reducerName] }
  );

const authenticationPersistFilter = whitelistInputWithInitialState(
  'authentication',
  ['landingRoute'],
  INITIAL_LOGIN_STATE
);

const namespacePersistFilter = whitelistInputWithInitialState(
  'namespaces',
  ['activeNamespaces'],
  INITIAL_NAMESPACE_STATE
);

const globalStateFilter = whitelistInputWithInitialState('globalState', ['language', 'theme'], INITIAL_GLOBAL_STATE);

const graphPersistFilter = whitelistInputWithInitialState('graph', ['filterState', 'layout'], INITIAL_GRAPH_STATE);

const userSettingsPersitFilter = whitelistInputWithInitialState(
  'userSettings',
  ['duration', 'refreshInterval', 'timeRange'],
  INITIAL_USER_SETTINGS_STATE
);

const persistConfig = {
  key: persistKey,
  storage: storage,
  whitelist: ['authentication', 'globalState', 'graph', 'namespaces', 'statusState', 'tracingState', 'userSettings'],
  transforms: [
    authenticationPersistFilter,
    graphPersistFilter,
    globalStateFilter,
    namespacePersistFilter,
    userSettingsPersitFilter
  ]
};

const composeEnhancers =
  (process.env.NODE_ENV === 'development' && window && window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__) || compose;

const configureStore = (initialState: KialiAppState): Store<KialiAppState, KialiAppAction> => {
  // configure middlewares
  const middlewares = [thunk];
  // compose enhancers
  const enhancer = composeEnhancers(applyMiddleware(...middlewares));
  // persist reducers
  const persistentReducer = persistReducer(persistConfig, rootReducer);

  return createStore(persistentReducer, initialState, enhancer);
};

// Setup the initial state of the Redux store with defaults
// (instead of having things be undefined until they are populated by query)
// Redux 4.0 actually required this
const initialStore: KialiAppState = {
  authentication: INITIAL_LOGIN_STATE,
  clusters: INITIAL_CLUSTER_STATE,
  globalState: INITIAL_GLOBAL_STATE,
  graph: INITIAL_GRAPH_STATE,
  istioStatus: INITIAL_ISTIO_STATUS_STATE,
  istioCertsInfo: INITIAL_ISTIO_CERTS_INFO_STATE,
  mesh: INITIAL_MESH_STATE,
  meshTLSStatus: INITIAL_MESH_TLS_STATE,
  metricsStats: INITIAL_METRICS_STATS_STATE,
  messageCenter: INITIAL_MESSAGE_CENTER_STATE,
  namespaces: INITIAL_NAMESPACE_STATE,
  statusState: INITIAL_STATUS_STATE,
  tourState: INITIAL_TOUR_STATE,
  tracingState: INITIAL_TRACING_STATE,
  userSettings: INITIAL_USER_SETTINGS_STATE
};

// pass an optional param to rehydrate state on app start
export const store = configureStore(initialStore);
export const persistor = persistStore(store);
