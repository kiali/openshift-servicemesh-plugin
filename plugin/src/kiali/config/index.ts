// Authentication

import { authenticationConfig } from './AuthenticationConfig';
// Configuration

import { config } from './Config';

// Icons
import { icons } from './Icons';

// Logos
import kialiLogoLight from '../assets/img/logo-lightbkg.svg';
import kialiLogoDark from '../assets/img/logo-darkbkg.svg';
import kialiIconLight from '../assets/img/icon-lightbkg.svg';
import kialiIconDark from '../assets/img/icon-darkbkg.svg';

// Paths
import { Paths } from './Paths';

// Jaeger Query
import { jaegerQuery } from './JaegerQuery';

// ServerConfig
import { homeCluster, isMultiCluster, serverConfig } from './ServerConfig';

export {
  authenticationConfig,
  config,
  Paths,
  icons,
  homeCluster,
  isMultiCluster,
  kialiLogoLight,
  kialiLogoDark,
  kialiIconLight,
  kialiIconDark,
  serverConfig,
  jaegerQuery
};
