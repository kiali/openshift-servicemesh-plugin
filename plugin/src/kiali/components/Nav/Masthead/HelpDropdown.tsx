import * as React from 'react';
import { AboutUIModal } from '../../About/AboutUIModal';
import { KialiAppState } from '../../../store/Store';
import { DebugInformation } from '../../../components/DebugInformation/DebugInformation';
import { QuestionCircleIcon } from '@patternfly/react-icons/';
import { connect } from 'react-redux';
import { isUpstream } from '../../UpstreamDetector/UpstreamDetector';
import { Status, ExternalServiceInfo, StatusKey } from '../../../types/StatusState';
import { config, serverConfig } from '../../../config';
import { IstioCertsInfo } from 'components/IstioCertsInfo/IstioCertsInfo';
import { kialiStyle } from 'styles/StyleUtils';
import { Dropdown, DropdownItem, DropdownList, MenuToggle, MenuToggleElement } from '@patternfly/react-core';

type HelpDropdownProps = {
  externalServices: ExternalServiceInfo[];
  status: Status;
  warningMessages: string[];
};

const menuToggleStyle = kialiStyle({
  marginTop: '0.25rem',
  verticalAlign: '-0.1rem'
});

const HelpDropdownComponent: React.FC<HelpDropdownProps> = (props: HelpDropdownProps) => {
  const [isDropdownOpen, setIsDropdownOpen] = React.useState<boolean>(false);
  const [isAboutModalOpen, setIsAboutModalOpen] = React.useState<boolean>(false);
  const [isDebugInformationOpen, setIsDebugInformationOpen] = React.useState<boolean>(false);
  const [isCertsInformationOpen, setIsCertsInformationOpen] = React.useState<boolean>(false);

  const onDropdownSelect = () => {
    setIsDropdownOpen(!isDropdownOpen);
  };

  const buildDocumentationLink = () => {
    const url = new URL(config.documentation.url);
    if (isUpstream) {
      const kialiCoreVersion = props.status[StatusKey.KIALI_CORE_VERSION] || 'unknown';

      url.searchParams.append('utm_source', 'kiali');
      url.searchParams.append('utm_medium', 'app');
      url.searchParams.append('utm_campaign', kialiCoreVersion);
      url.searchParams.append('utm_content', '?-menu');
    }
    return url.toString();
  };

  const items: JSX.Element[] = [];

  items.push(
    <DropdownItem key={'view_documentation'} onClick={() => window.open(buildDocumentationLink(), '_blank')}>
      Documentation
    </DropdownItem>
  );

  items.push(
    <DropdownItem key={'view_debug_info'} onClick={() => setIsDebugInformationOpen(true)}>
      View Debug Info
    </DropdownItem>
  );

  if (serverConfig.kialiFeatureFlags.certificatesInformationIndicators.enabled) {
    items.push(
      <DropdownItem key={'view_certs_info'} onClick={() => setIsCertsInformationOpen(true)}>
        View Certificates Info
      </DropdownItem>
    );
  }

  items.push(
    <DropdownItem key={'view_about_info'} onClick={() => setIsAboutModalOpen(true)}>
      About
    </DropdownItem>
  );

  return (
    <>
      <AboutUIModal
        status={props.status}
        externalServices={props.externalServices}
        warningMessages={props.warningMessages}
        isOpen={isAboutModalOpen}
        onClose={() => setIsAboutModalOpen(false)}
      />

      <DebugInformation isOpen={isDebugInformationOpen} onClose={() => setIsDebugInformationOpen(false)} />

      {serverConfig.kialiFeatureFlags.certificatesInformationIndicators.enabled && (
        <IstioCertsInfo isOpen={isCertsInformationOpen} onClose={() => setIsCertsInformationOpen(false)} />
      )}

      <Dropdown
        toggle={(toggleRef: React.Ref<MenuToggleElement>) => (
          <MenuToggle
            ref={toggleRef}
            className={menuToggleStyle}
            data-test="about-help-button"
            onClick={() => setIsDropdownOpen(!isDropdownOpen)}
            aria-label="Help"
            variant="plain"
            isExpanded={isDropdownOpen}
          >
            <QuestionCircleIcon />
          </MenuToggle>
        )}
        isOpen={isDropdownOpen}
        popperProps={{ position: 'right' }}
        onOpenChange={(isOpen: boolean) => setIsDropdownOpen(isOpen)}
        onSelect={onDropdownSelect}
      >
        <DropdownList>{items}</DropdownList>
      </Dropdown>
    </>
  );
};

const mapStateToProps = (state: KialiAppState) => ({
  status: state.statusState.status,
  externalServices: state.statusState.externalServices,
  warningMessages: state.statusState.warningMessages
});

export const HelpDropdown = connect(mapStateToProps)(HelpDropdownComponent);
