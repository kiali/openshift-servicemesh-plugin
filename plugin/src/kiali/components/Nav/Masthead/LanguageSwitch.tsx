import * as React from 'react';
import { kialiStyle } from 'styles/StyleUtils';
import { Dropdown, DropdownItem, DropdownList, MenuToggle, MenuToggleElement, Tooltip } from '@patternfly/react-core';
import { KialiIcon } from 'config/KialiIcon';
import { GlobalActions } from 'actions/GlobalActions';
import { store } from 'store/ConfigStore';
import { I18N_NAMESPACE, Language } from 'types/Common';
import { KialiAppState } from 'store/Store';
import { connect } from 'react-redux';
import { useTranslation } from 'react-i18next';
import { i18n } from 'i18n';

const menuToggleStyle = kialiStyle({
  marginTop: '0.25rem',
  verticalAlign: '-0.125rem'
});

const iconStyle = kialiStyle({
  verticalAlign: '-0.375rem',
  $nest: {
    '& svg': {
      width: '1.5rem',
      height: '1.5rem'
    }
  }
});

const checkStyle = kialiStyle({
  marginLeft: '0.5rem'
});

type LanguageSwitchProps = {
  language: string;
};

export const LanguageSwitchComponent: React.FC<LanguageSwitchProps> = ({ language }) => {
  const [isDropdownOpen, setIsDropdownOpen] = React.useState<boolean>(false);

  const { t } = useTranslation(I18N_NAMESPACE);

  const languageItems: React.ReactNode[] = [
    <DropdownItem key="English" onClick={() => switchLanguage(Language.ENGLISH)}>
      <span>English</span>
      {language === Language.ENGLISH && <KialiIcon.Check className={checkStyle} />}
    </DropdownItem>,

    <DropdownItem key="Chinese" onClick={() => switchLanguage(Language.CHINESE)}>
      <span>中文</span>
      {language === Language.CHINESE && <KialiIcon.Check className={checkStyle} />}
    </DropdownItem>
  ];

  const switchLanguage = (language: string): void => {
    i18n.changeLanguage(language);
    store.dispatch(GlobalActions.setLanguage(language));
  };

  const onDropdownSelect = (): void => {
    setIsDropdownOpen(!isDropdownOpen);
  };

  return (
    <Tooltip position="bottom" content={<>{t('Switch language')}</>} trigger="mouseenter click" exitDelay={0}>
      <Dropdown
        toggle={(toggleRef: React.Ref<MenuToggleElement>) => (
          <MenuToggle
            ref={toggleRef}
            className={menuToggleStyle}
            data-test="switch-language-button"
            onClick={() => setIsDropdownOpen(!isDropdownOpen)}
            aria-label="Switch language"
            variant="plain"
            isExpanded={isDropdownOpen}
          >
            <KialiIcon.Language className={iconStyle} />
          </MenuToggle>
        )}
        isOpen={isDropdownOpen}
        popperProps={{ position: 'center' }}
        onOpenChange={(isOpen: boolean) => setIsDropdownOpen(isOpen)}
        onSelect={onDropdownSelect}
      >
        <DropdownList>{languageItems}</DropdownList>
      </Dropdown>
    </Tooltip>
  );
};

const mapStateToProps = (state: KialiAppState): LanguageSwitchProps => {
  return {
    language: state.globalState.language
  };
};

export const LanguageSwitch = connect(mapStateToProps)(LanguageSwitchComponent);
