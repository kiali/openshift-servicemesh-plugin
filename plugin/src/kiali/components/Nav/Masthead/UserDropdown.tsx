import * as React from 'react';
import { SessionTimeout } from '../../SessionTimeout/SessionTimeout';
import { config } from '../../../config';
import { MILLISECONDS } from '../../../types/Common';
import { Timer } from 'globals';
import { KialiAppState, LoginSession } from '../../../store/Store';
import { authenticationConfig } from '../../../config/AuthenticationConfig';
import { AuthStrategy } from '../../../types/Auth';
import moment from 'moment';
import { KialiDispatch } from 'types/Redux';
import { LoginThunkActions } from '../../../actions/LoginThunkActions';
import { connect } from 'react-redux';
import * as API from '../../../services/Api';
import { kialiStyle } from 'styles/StyleUtils';
import { Dropdown, DropdownItem, DropdownList, MenuToggle, MenuToggleElement } from '@patternfly/react-core';

type UserProps = {
  extendSession: (session: LoginSession) => void;
  logout: () => void;
  session?: LoginSession;
};

type UserState = {
  checkSessionTimerId?: Timer;
  isDropdownOpen: boolean;
  isSessionTimeoutDismissed: boolean;
  showSessionTimeOut: boolean;
  timeCountDownSeconds: number;
  timeLeftTimerId?: Timer;
};

const dropdownStyle = kialiStyle({
  paddingLeft: 0,
  paddingRight: 0
});

class UserDropdownComponent extends React.Component<UserProps, UserState> {
  constructor(props: UserProps) {
    super(props);

    this.state = {
      showSessionTimeOut: false,
      timeCountDownSeconds: this.timeLeft() / MILLISECONDS,
      isSessionTimeoutDismissed: false,
      isDropdownOpen: false
    };
  }

  componentDidMount() {
    const checkSessionTimerId = setInterval(() => {
      this.checkSession();
    }, 3000);

    const timeLeftTimerId = setInterval(() => {
      this.setState({ timeCountDownSeconds: this.timeLeft() / MILLISECONDS });
    }, 1000);

    this.setState({
      checkSessionTimerId: checkSessionTimerId,
      timeLeftTimerId: timeLeftTimerId
    });
  }

  componentWillUnmount() {
    if (this.state.checkSessionTimerId) {
      clearInterval(this.state.checkSessionTimerId);
    }

    if (this.state.timeLeftTimerId) {
      clearInterval(this.state.timeLeftTimerId);
    }
  }

  timeLeft = (): number => {
    const expiresOn = moment(this.props.session!.expiresOn);

    if (expiresOn <= moment()) {
      this.props.logout();
    }

    return expiresOn.diff(moment());
  };

  checkSession = () => {
    if (this.timeLeft() < config.session.timeOutforWarningUser) {
      this.setState({ showSessionTimeOut: true });
    }
  };

  handleLogout = () => {
    if (authenticationConfig.logoutEndpoint) {
      API.logout()
        .then(_ => {
          (document.getElementById('openshiftlogout') as HTMLFormElement).submit();
        })
        .catch(error => {
          const errorMsg = error.response && error.response.data.error ? error.response.data.error : error.message;
          console.error(`Logout failed. "kiali-token" cookie may need to be cleared manually: ${errorMsg}`);
          (document.getElementById('openshiftlogout') as HTMLFormElement).submit();
        });
    } else {
      this.props.logout();
    }
  };

  extendSession = (session: LoginSession) => {
    this.props.extendSession(session);
    this.setState({ showSessionTimeOut: false });
  };

  onDropdownToggle = isDropdownOpen => {
    this.setState({
      isDropdownOpen
    });
  };

  onDropdownSelect = _event => {
    this.setState({
      isDropdownOpen: !this.state.isDropdownOpen
    });
  };

  render() {
    const { isDropdownOpen } = this.state;

    const canLogout =
      authenticationConfig.strategy !== AuthStrategy.anonymous && authenticationConfig.strategy !== AuthStrategy.header;

    const userDropdownItems = (
      <DropdownItem key={'user_logout_option'} onClick={this.handleLogout} isDisabled={!canLogout}>
        Logout
      </DropdownItem>
    );

    return (
      <>
        <SessionTimeout
          onLogout={this.props.logout}
          onExtendSession={this.extendSession}
          onDismiss={this.dismissHandler}
          show={this.state.showSessionTimeOut && !this.state.isSessionTimeoutDismissed}
          timeOutCountDown={this.state.timeCountDownSeconds}
        />

        {this.props.session && !canLogout && <>{this.props.session.username}</>}

        {this.props.session && canLogout && (
          <Dropdown
            toggle={(toggleRef: React.Ref<MenuToggleElement>) => (
              <MenuToggle
                ref={toggleRef}
                className={dropdownStyle}
                variant="plainText"
                onClick={() => this.onDropdownToggle(!isDropdownOpen)}
                isExpanded={isDropdownOpen}
              >
                {this.props.session?.username}
              </MenuToggle>
            )}
            isOpen={isDropdownOpen}
            onSelect={this.onDropdownSelect}
            popperProps={{ position: 'right' }}
            onOpenChange={(isOpen: boolean) => this.onDropdownToggle(isOpen)}
          >
            <DropdownList>{[userDropdownItems]}</DropdownList>
          </Dropdown>
        )}

        {authenticationConfig.strategy === AuthStrategy.openshift && authenticationConfig.logoutEndpoint && (
          <form id="openshiftlogout" action={authenticationConfig.logoutEndpoint} method="post">
            <input type="hidden" name="then" value={authenticationConfig.logoutRedirect} />
          </form>
        )}
      </>
    );
  }

  private dismissHandler = () => {
    this.setState({ isSessionTimeoutDismissed: true });
  };
}

const mapStateToProps = (state: KialiAppState) => ({
  session: state.authentication.session
});

const mapDispatchToProps = (dispatch: KialiDispatch) => ({
  logout: () => dispatch(LoginThunkActions.logout()),
  extendSession: (session: LoginSession) => dispatch(LoginThunkActions.extendSession(session))
});

export const UserDropdown = connect(mapStateToProps, mapDispatchToProps)(UserDropdownComponent);
