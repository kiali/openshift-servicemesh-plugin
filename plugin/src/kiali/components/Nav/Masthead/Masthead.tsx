import * as React from 'react';
import { Label, Flex, FlexItem, Tooltip, Toolbar, ToolbarItem } from '@patternfly/react-core';
import { ClusterIcon } from '@patternfly/react-icons';

import { serverConfig } from '../../../config';
import { default as MeshMTLSStatus } from '../../../components/MTls/MeshMTLSStatus';
import { default as IstioStatus } from '../../IstioStatus/IstioStatus';
import PfSpinner from '../../PfSpinner';
import UserDropdown from './UserDropdown';
import HelpDropdown from './HelpDropdown';
import MessageCenterTriggerContainer from '../../../components/MessageCenter/MessageCenterTrigger';

class MastheadItems extends React.Component {
  render() {
    return (
      <>
        <PfSpinner />
        <Toolbar>
          <ToolbarItem>
            <Flex>
              <FlexItem align={{ default: 'alignRight' }}>
                {!!serverConfig.clusterInfo?.name && (
                  <Tooltip
                    entryDelay={0}
                    position="bottom"
                    content={<div>Kiali home cluster: {serverConfig.clusterInfo.name}</div>}
                  >
                    <Label color="blue" icon={<ClusterIcon />}>
                      {serverConfig.clusterInfo.name}
                    </Label>
                  </Tooltip>
                )}
              </FlexItem>
              <FlexItem>
                <IstioStatus cluster={serverConfig.clusterInfo?.name} />
              </FlexItem>
              <FlexItem>
                <MeshMTLSStatus />
              </FlexItem>
              <FlexItem style={{ marginRight: 0 }}>
                <MessageCenterTriggerContainer />
              </FlexItem>
              <FlexItem>
                <HelpDropdown />
              </FlexItem>
              <FlexItem>
                <UserDropdown />
              </FlexItem>
            </Flex>
          </ToolbarItem>
        </Toolbar>
      </>
    );
  }
}

export default MastheadItems;
