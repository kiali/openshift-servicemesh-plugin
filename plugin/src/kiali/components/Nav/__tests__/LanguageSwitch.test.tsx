import * as React from 'react';
import { shallowToJson } from 'enzyme-to-json';
import { mount, shallow } from 'enzyme';
import { LanguageSwitchComponent } from '../Masthead/LanguageSwitch';
import { MenuToggle } from '@patternfly/react-core';
import { store } from 'store/ConfigStore';
import { Language } from 'types/Common';
import { serverConfig, setServerConfig } from 'config/ServerConfig';

const i18nServerConfig = Object.assign({}, serverConfig);

describe('Language switch', () => {
  beforeAll(() => {
    setServerConfig(i18nServerConfig);
  });
  it('renders correctly', () => {
    const wrapper = shallow(<LanguageSwitchComponent language={Language.ENGLISH} />);

    expect(shallowToJson(wrapper)).toBeDefined();
    expect(shallowToJson(wrapper)).toMatchSnapshot();
  });

  it('changes to english language', () => {
    const wrapper = mount(<LanguageSwitchComponent language={Language.CHINESE} />);

    // click menu toggle
    wrapper.find(MenuToggle).simulate('click');

    // select English option
    wrapper
      .findWhere(node => node.key() === 'English')
      .findWhere(node => node.type() === 'button')
      .simulate('click');

    expect(store.getState().globalState.language).toBe(Language.ENGLISH);
  });

  it('changes to chinese language', () => {
    const wrapper = mount(<LanguageSwitchComponent language={Language.ENGLISH} />);

    // click menu toggle
    wrapper.find(MenuToggle).simulate('click');

    // select Chinese option
    wrapper
      .findWhere(node => node.key() === 'Chinese')
      .findWhere(node => node.type() === 'button')
      .simulate('click');

    expect(store.getState().globalState.language).toBe(Language.CHINESE);
  });
});
