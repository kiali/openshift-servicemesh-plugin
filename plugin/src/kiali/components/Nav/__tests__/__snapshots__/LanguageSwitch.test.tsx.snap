// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Language switch renders correctly 1`] = `
<Tooltip
  content={
    <React.Fragment>
      Switch language
    </React.Fragment>
  }
  exitDelay={0}
  position="bottom"
  trigger="mouseenter click"
>
  <Dropdown
    isOpen={false}
    onOpenChange={[Function]}
    onSelect={[Function]}
    popperProps={
      Object {
        "position": "center",
      }
    }
    toggle={[Function]}
  >
    <DropdownList>
      <DropdownItem
        key="English"
        onClick={[Function]}
      >
        <span>
          English
        </span>
        <Check
          className="kiali_fm5q8n4"
        />
      </DropdownItem>
      <DropdownItem
        key="Chinese"
        onClick={[Function]}
      >
        <span>
          中文
        </span>
      </DropdownItem>
    </DropdownList>
  </Dropdown>
</Tooltip>
`;
